### ISSUE TITLE
Xbox One S controller on new firmware has incorrect mapping

### ISSUE THREAD

**Issue by XirXes** [atar-axis/xpadneo#314]
## Controller Model

1708 Firmware 5.11.3116.0

* [x] Xbox One S controller
* [ ] Xbox Elite 2 controller
* [ ] Xbox Series X|S controller
* [ ] Other:

### Connection mode

* [x] Bluetooth connection
* [ ] Xbox Dongle connection (not yet supported)
* [ ] USB cable (not yet supported)

## Installed Software

* [x] Steam Input (enabled by default via Steam Desktop client)
* [ ] Steam Link (usually via Raspberry Pi or other micro computers)
* [X] xow (alternative driver using the Xbox dongle)
* [ ] Anti-Micro (may affect button mappings)
* [ ] netstick (shares input devices via network similar to Steam Link)
* [ ] xboxdrv (user-space gamepad driver)

## Severity / Impact

* [x] I've read the docs and the bug reporting instructions
* [x] I've applied the latest firmware update to the controller
* [x] It does not work at all
* [ ] It used to work in a previous version
* [ ] It mostly works but sometimes it doesn't
* [ ] I found a work-around
* [ ] I probably didn't figure it all out but it's too early to give up
* [ ] I don't know how to ... 
* [ ] It's too complicated
* [ ] Fantastic work but ... 
* [ ] I can code and I want to help

## Describe the Bug

This controller is running the new beta firmware with double press to toggle between bluetooth and xbox wireless. I believe the firmware version matches the Series controllers; 5.11.3116.0
The controller connects to bluetooth fine and is detected by games and Steam, but the layout is wrong and some buttons don't work. It doesn't do the vibration that xpadneo usually does, I don't think it connects to xpadneo at all, its descriptor is always under `/sys/module/hid_generic/drivers/hid\:hid-generic/0005:045E:*/report_descriptor`. 

## System Information

```console
# uname -a
Linux nexus 5.14.11-xanmod1-1 #1 SMP PREEMPT Mon, 11 Oct 2021 01:35:41 +0000 x86_64 GNU/Linux
```

```console
# xxd -c20 -g1 /sys/module/hid_xpadneo/drivers/hid:xpadneo/0005:045E:*/report_descriptor | tee >(cksum)

$ xxd -c20 -g1 /sys/module/hid_generic/drivers/hid\:hid-generic/0005:045E:*/report_descriptor | tee >(cksum)
00000000: 05 01 09 05 a1 01 85 01 09 01 a1 00 09 30 09 31 15 00 27 ff  .............0.1..'.
00000014: ff 00 00 95 02 75 10 81 02 c0 09 01 a1 00 09 32 09 35 15 00  .....u.........2.5..
00000028: 27 ff ff 00 00 95 02 75 10 81 02 c0 05 02 09 c5 15 00 26 ff  '......u..........&.
0000003c: 03 95 01 75 0a 81 02 15 00 25 00 75 06 95 01 81 03 05 02 09  ...u.....%.u........
00000050: c4 15 00 26 ff 03 95 01 75 0a 81 02 15 00 25 00 75 06 95 01  ...&....u.....%.u...
00000064: 81 03 05 01 09 39 15 01 25 08 35 00 46 3b 01 66 14 00 75 04  .....9..%.5.F;.f..u.
00000078: 95 01 81 42 75 04 95 01 15 00 25 00 35 00 45 00 65 00 81 03  ...Bu.....%.5.E.e...
0000008c: 05 09 19 01 29 0f 15 00 25 01 75 01 95 0f 81 02 15 00 25 00  ....)...%.u.......%.
000000a0: 75 01 95 01 81 03 05 0c 0a b2 00 15 00 25 01 95 01 75 01 81  u............%...u..
000000b4: 02 15 00 25 00 75 07 95 01 81 03 05 0f 09 21 85 03 a1 02 09  ...%.u........!.....
000000c8: 97 15 00 25 01 75 04 95 01 91 02 15 00 25 00 75 04 95 01 91  ...%.u.......%.u....
000000dc: 03 09 70 15 00 25 64 75 08 95 04 91 02 09 50 66 01 10 55 0e  ..p..%du......Pf..U.
000000f0: 15 00 26 ff 00 75 08 95 01 91 02 09 a7 15 00 26 ff 00 75 08  ..&..u.........&..u.
00000104: 95 01 91 02 65 00 55 00 09 7c 15 00 26 ff 00 75 08 95 01 91  ....e.U..|..&..u....
00000118: 02 c0 c0

```

## Controller and Bluetooth Information

0a5c:21e8 Broadcom Corp. BCM20702A0 Bluetooth 4.0

[xpadneo-btmon.txt](https://github.com/atar-axis/xpadneo/files/7325971/xpadneo-btmon.txt)
[xpadneo-dmesg.txt](https://github.com/atar-axis/xpadneo/files/7325972/xpadneo-dmesg.txt)
[xpadneo-lsusb.txt](https://github.com/atar-axis/xpadneo/files/7325974/xpadneo-lsusb.txt)

**Comment by kakra** [comment]
This looks at least like it should be able to be detected and used by xpadneo:
```dmesg
[33449.020866] hid-generic 0005:045E:0B20.0024: input,hidraw0: BLUETOOTH HID v5.11 Gamepad [Xbox Wireless Controller] on 5c:f3:70:64:14:54
[33632.964263] input: Xbox One Wireless Controller as /devices/virtual/input/input80
[33633.679035] input: Microsoft X-Box 360 pad as /devices/virtual/input/input81
[33651.325910] input: Xbox Wireless Controller as /devices/virtual/misc/uhid/0005:045E:0B20.0025/input/input82
[33651.326237] hid-generic 0005:045E:0B20.0025: input,hidraw0: BLUETOOTH HID v5.11 Gamepad [Xbox Wireless Controller] on 5c:f3:70:64:14:54
```

So what does it tell us:

* The new firmware comes with a new product ID 0x0B20. I can simply add that to xpadneo.
* I'm not sure why "Microsoft X-Box 360 pad" is showing up. Did you connect it via USB somehow? Usually, only the xpad driver uses that name - and that only supports USB. But it shows up on the virtual bus so it is injected by bluez via uhid/uinput.
* There's a delay of about 20s between those detections. That could either mean that pairing/connecting took a while and had some hiccups, or that you connected and disconnected the USB cable.
* Is this the MAC address of your controller? C8:3F:26:87:DF:D5

**Comment by XirXes** [comment]
I think the 360 pad is Steam input, I may not have closed Steam completely when testing, and I run with Xbox controller configuration enabled. Either that or XOW masking the controller, disregard it for sure. I had the rf paired and was toggling the controller mode to connect to bluetooth.
That is the MAC address of the controller, just confirmed here. I'll be happy to run a build here to test, but it may be easiest to test with the changes actually pushed to the repo, I'd just have to rebuild the dkms package in that case.

**Comment by XirXes** [comment]
My controller just got updated to 5.11.3118.0. I'll run some more tests tonight.

**Comment by kakra** [comment]
Thanks for the heads up.

**Comment by kakra** [comment]
@XirXes Can you retest with the latest PR #315 which should fix SDL2 mapping fixups for PID 0xB20?

**Comment by XirXes** [comment]
I'm sorry I didn't see this earlier, I just tested with the updated changes to PR #315 and everything but the 'View' button is working perfectly. It isn't making any event under either device.

**Comment by kakra** [comment]
What's the view button? The left one with the two windows icon?

**Comment by XirXes** [comment]
Yeah, the select button, my guess is its getting filtered, its HID ID was 9000b, which collides with the Xbox logo button. Line 125 in `hid-xpadneo.c` mentions it

**Comment by kakra** [comment]
This should not have been auto-closed, reopening...

> Yeah, the select button, my guess is its getting filtered, its HID ID was 9000b, which collides with the Xbox logo button. Line 125 in `hid-xpadneo.c` mentions it.

Please check if both thumb stick buttons act correctly. If not, the buttons are shifted and a wrong mapping is assumed somewhere. If they do work correctly, we may need some model quirk which handles this single button.

**Comment by XirXes** [comment]
Just double checked, the thumbsticks are both working correctly. The mode button is also getting mapped properly to the Xbox Wireless Controller Consumer Control event device. Only the select button not registering is left.

**Comment by kakra** [comment]
Are you able to create an annotated btmon log of each button press?

**Comment by XirXes** [comment]
Not until this evening, but I'd be happy to.

**Comment by kakra** [comment]
Ok, thanks. Probably exclude the sticks: We know those are correct, and you won't be able to click them without moving them, which creates a lot of useless noise in the log. Unless that's your secret superpower... ;-)

**Comment by XirXes** [comment]
[xpadneo-btmon-211116-1.txt](https://github.com/atar-axis/xpadneo/files/7551402/xpadneo-btmon-211116-1.txt)

Timestamps
29  A
41  B
56  X
70  Y
96  LB
106  RB
184  Select
198  Start
212  Mode

**Comment by XirXes** [comment]
Verified controller model 1914 is working perfectly on the build of xpadneo I'm running. The elite series 2 also connects and works correctly, with mapping correct, but the mode button and the profile button both register under the base xbox wireless controller, and its no longer listed as an xbox elite controller, evtest just showed Xbox Wireless Controller and Xbox Wireless Controller Consumer Control.

**Comment by XirXes** [comment]
I made an updated fork of the [firmware/xb1s-0b20 branch](https://github.com/XirXes/xpadneo/tree/firmware/xb1s-0b20) with the latest changes from master, and confirmed every button is mapped correctly on 1708 running 5.11.3118.0 with select(view) still being filtered and not functioning. 

1914 running 5.11.3118.0 has no issues with keymaps.

**Comment by XirXes** [comment]
Did not mean to close this issue, the select button still doesn't register on 1708 on firmware 5.11.3118.0

**Comment by XirXes** [comment]
Did not mean to close this, select button still doesn't register on model 1708 on firmware 5.11.3118.0.

**Comment by kakra** [comment]
Which is the 1708 model? The original XB1S controller?

**Comment by XirXes** [comment]
1708 is the XB1S, correct.

I will know later today if its on Windows. The biggest problem I have is all of my controllers are up to date through the xbox. I might be able to downgrade one of them.

**Comment by tmg0mes** [comment]
Hey, any news on the select button issue on 1708 model?

**Comment by kakra** [comment]
Please check again with latest master.

**Comment by XirXes** [comment]
I spotted a Reddit thread saying they've updated Xbox Accessories in Windows 10 and 11 and the new update includes firmware 5.13.3143. [link](https://www.reddit.com/r/xboxinsiders/comments/ta7kib/xbox_one_controller_firmware_5133143_available/)

As for retesting with a new build, I just did and I still don't get any response from btn_select, and it does enable the record button even though this controller doesn't have one.

```
Input device ID: bus 0x5 vendor 0x45e product 0x2e0 version 0x903
Input device name: "Xbox Wireless Controller Consumer Control"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 167 (KEY_RECORD)
    Event code 373 (KEY_MODE)
```

**Comment by kakra** [comment]
Thanks for spotting this. I tried all my models and only the XBXS was updated to 5.13. My XB1S and XBE2 remained at 4.8. The updates probably roll out in waves.

Regarding the buttons, xpadneo will currently report all buttons available across all models even if that particular model does not have each button.

All buttons worked as they should. I'll probably have to wait for my older models to update to 5.13.

**Comment by ryannathans** [comment]
> I spotted a Reddit thread saying they've updated Xbox Accessories in Windows 10 and 11 and the new update includes firmware 5.13.3143. [link](https://www.reddit.com/r/xboxinsiders/comments/ta7kib/xbox_one_controller_firmware_5133143_available/)
> 
> As for retesting with a new build, I just did and I still don't get any response from btn_select, and it does enable the record button even though this controller doesn't have one.
> 
> ```
> Input device ID: bus 0x5 vendor 0x45e product 0x2e0 version 0x903
> Input device name: "Xbox Wireless Controller Consumer Control"
> Supported events:
>   Event type 0 (EV_SYN)
>   Event type 1 (EV_KEY)
>     Event code 167 (KEY_RECORD)
>     Event code 373 (KEY_MODE)
> ```

for me it seems select and share are swapped in the latest master
issues doesn't exist in the last release

**Comment by XirXes** [comment]
> Thanks for spotting this. I tried all my models and only the XBXS was updated to 5.13. My XB1S and XBE2 remained at 4.8. The updates probably roll out in waves.
> 
> Regarding the buttons, xpadneo will currently report all buttons available across all models even if that particular model does not have each button.
> 
> All buttons worked as they should. I'll probably have to wait for my older models to update to 5.13.

I have confirmed what @ryannathans said, there was a second update to the Xbox Accessories app and it does include firmware 5.13.3143 for the 1708 model controllers, so you should be able to get your One S controllers updated now.

The only big note about the controller update in my eyes is if you need non LE bluetooth on your controller for anything, you are best off staying on or reverting to the old firmware.

You can also now downgrade controllers from Windows by entering this URL in the Run dialog box;
```
xboxaccessories:\\firmwareupdate?legacyDowngrade=true
```

**Comment by kakra** [comment]
> I have confirmed what @ryannathans said, there was a second update to the Xbox Accessories app and it does include firmware 5.13.3143 for the 1708 model controllers, so you should be able to get your One S controllers updated now.

I can confirm that for all my three models. Although the app says "no update available" it will still apply an update to 5.13 when I click the update button. I can now test this stuff and already fixed two problems in the source code but currently my XBE2 controller no longer connects to Bluetooth and just says:
```
[275473.672782] Bluetooth: hci0: unexpected SMP command 0x08 from 98:7a:14:61:17:4d
[275473.673563] Bluetooth: hci0: unexpected SMP command 0x09 from 98:7a:14:61:17:4d
```

It then semi-connects (xpadneo sees the device) but it won't work and eventually timeout:
```
[275294.926036] Bluetooth: hci0: No such LE device 98:7a:14:61:17:4d (0x0)
```

It seems that with version 5.13, both models XB1S and XBXS use the same PID now (0B13). The button mapping is correct for both models.

**Comment by XirXes** [comment]
> It seems that with version 5.13, both models XB1S and XBXS use the same PID now (0B13). The button mapping is correct for both models.

I can confirm that with the latest commit [4fd620c](https://github.com/atar-axis/xpadneo/commit/4fd620cd6cb80fb0c1490dc1c864108679d91ab1) and modprobing hid-xpadneo before I connect the controller my select button now registers! With that the keymap issues I was having on 5.13 are completely resolved. I'll close this issue, now that you have the new firmware in hand I'm sure if there's anything else wrong you'll find it.
